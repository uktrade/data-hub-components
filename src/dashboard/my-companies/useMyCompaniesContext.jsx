import { orderBy, pick } from 'lodash'
import { useReducer } from 'react'
import createUseContext from 'constate'
import { FILTER_CHANGE, LIST_CHANGE, ORDER_CHANGE } from './constants'

export const filterCompanyName = (companies, filterText) =>
  filterText.length
    ? companies.filter((c) =>
        c.company.name.toLowerCase().includes(filterText.toLowerCase())
      )
    : companies

const defaultState = {
  lists: [],
  selectedIdx: 0,
  sortBy: 'recent',
  filter: '',
}

export const reducer = (state, { type, ...action }) => {
  switch (type) {
    case LIST_CHANGE:
      return { ...state, selectedIdx: action.idx }
    case FILTER_CHANGE:
      return { ...state, filter: action.filter }
    case ORDER_CHANGE:
      return { ...state, sortBy: action.sortBy }
    default:
      return state
  }
}

// We are unpacking children here just to remove them from initialState.
const useMyCompaniesContext = createUseContext(
  ({ children, editListsLinkProps, ...rest }) => {
    const initialState = pick(rest, [
      'lists',
      'selectedIdx',
      'sortBy',
      'filter',
    ])
    const [state, dispatch] = useReducer(reducer, {
      ...defaultState,
      ...initialState,
      // Reducer assumes the list of company lists is sorted.
      lists: orderBy(initialState.lists, 'name'),
    })

    return {
      state,
      dispatch,
      editListsLinkProps,
    }
  }
)

export default useMyCompaniesContext
