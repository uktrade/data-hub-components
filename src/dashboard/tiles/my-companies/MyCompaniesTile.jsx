import React, { useState } from 'react'

import MyCompaniesFilters from './MyCompaniesFilters'

function MyCompaniesTile({ data }) {
  const [companies, setCompanies] = useState(data)

  function onFilterUpdate(newCompanies) {
    setCompanies(newCompanies)
  }

  return (
    <>
      <MyCompaniesFilters initialCompanies={data} companies={companies} onFilterUpdate={onFilterUpdate} />
      <pre>{JSON.stringify(companies)}</pre>
    </>
  )
}

export default MyCompaniesTile
