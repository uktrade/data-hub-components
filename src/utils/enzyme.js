export const withTargetValue = value => ({
  target: {
    value,
  },
})

/**
 * @function getOptionValueByText
 * @description Finds an `<option>` element with its inner text matching the
 * {text} parameter and returns the value of its `value` attribute.
 * @param {*} wrapper - Enzyme React wrapper
 * @param {String} text - The option text
 */
export const getOptionValueByText = (wrapper, text) => (
  wrapper.find('option')
    .find({ children: text })
    .props()
    .value
)

/**
 * @function changeAndUpdate
 * @description Simulates the `change` event on the provided {wrapper} with
 * {vaule} as `target.value` and updates (re-renders) the wrapper.
 * @param {*} wrapper - Enzyme React wrapper
 * @param {*} value - The value to be used as `event.target`
 */
export const changeAndUpdate = (wrapper, value) => {
  wrapper.simulate('change', withTargetValue(value))
  wrapper.update()
}

/**
 * @function changeSelectAndUpdate
 * @description Finds a `<select>` element in the wrapper, selects the option
 * matching the {text} and updates (re-renders) the wrapper.
 * @param {*} wrapper - Enzyme React wrapper
 * @param {String} text - The option text
 */
export const changeSelectAndUpdate = (wrapper, text) => {
  changeAndUpdate(wrapper.find('select'), getOptionValueByText(wrapper, text))
  wrapper.update()
}
