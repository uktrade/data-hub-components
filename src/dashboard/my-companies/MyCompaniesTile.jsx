import React, { Fragment } from 'react'
import Details from '@govuk-react/details'
import useMyCompaniesContext from './useMyCompaniesContext'

import DashboardSection from '../DashboardSection'
import MyCompaniesFilters from './MyCompaniesFilters'
import MyCompaniesTable from './MyCompaniesTable'

const HelpText = () => (
  <Details summary="How do I remove a company from this list">
    You can remove a company from your list in the company&apos;s page.
  </Details>
)

function MyCompaniesTile() {
  const { state } = useMyCompaniesContext()

  const hasCompanies = !!state.companies.length
  const isFiltering = !!state.filterText.length
  const hasNoCompaniesInWatchList = !hasCompanies && !isFiltering
  const hasNoResults = !!state.companiesInitial.length && !state.companies.length

  const SubHeading = (
    <Fragment>
      You have not added any companies to your list. <br />
      You can add companies to this list from a company page, and only you can see this list.
    </Fragment>
  )

  return (
    <DashboardSection
      heading="My Companies"
      headingSlotComponent={!!state.companiesInitial.length && <MyCompaniesFilters />}
      subHeading={SubHeading}
      showSubHeading={hasNoCompaniesInWatchList}
    >
      {hasCompanies && (
        <Fragment>
          <MyCompaniesTable /> <HelpText />
        </Fragment>
      )}
      {hasNoResults && <p>No companies match your search</p>}
    </DashboardSection>
  )
}

export default MyCompaniesTile
