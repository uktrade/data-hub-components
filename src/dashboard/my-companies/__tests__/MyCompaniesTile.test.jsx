import React from 'react'
import { mount } from 'enzyme'
import MyCompaniesTile from '../MyCompaniesTile'
import useMyCompaniesContext from '../useMyCompaniesContext'
import companies from '../../__fixtures__/companies'

describe('My companies dashboard', () => {
  test('should match the snapshot', () => {
    const wrapper = mount(
      <useMyCompaniesContext.Provider mockInitialState={{ companiesInitial: companies, companies }}>
        <MyCompaniesTile />
      </useMyCompaniesContext.Provider>,
    )
    expect(wrapper).toMatchSnapshot()
  })
  describe('when there are a list of companies', () => {
    const wrapper = mount(
      <useMyCompaniesContext.Provider mockInitialState={{ companiesInitial: companies, companies }}>
        <MyCompaniesTile />
      </useMyCompaniesContext.Provider>,
    )
    test('should render filters', () => {
      expect(wrapper.find('input[type="text"]')).toHaveLength(1)
      expect(wrapper.find('select')).toHaveLength(1)
    })
    test('should render a table', () => {
      expect(wrapper.find('table')).toHaveLength(1)
    })
    test('should render help text', () => {
      expect(wrapper.find('details')).toHaveLength(1)
    })
  })
  describe('when there are no companies', () => {
    const wrapper = mount(
      <useMyCompaniesContext.Provider data={[]}>
        <MyCompaniesTile />
      </useMyCompaniesContext.Provider>,
    )
    test('should not render filters', () => {
      expect(wrapper.find('input[type="text"]')).toHaveLength(0)
      expect(wrapper.find('select')).toHaveLength(0)
    })
    test('should not render a table', () => {
      expect(wrapper.find('table')).toHaveLength(0)
    })
    test('should not render help text', () => {
      expect(wrapper.find('details')).toHaveLength(0)
    })
    test('should render no companies message', () => {
      expect(wrapper.find('p').text()).toContain('You have not added any companies to your list. You can add companies to this list from a company page, and only you can see this list.')
    })
  })
  describe('Search for companies', () => {
    test('should filter a search result', () => {
      const wrapper = mount(
        <useMyCompaniesContext.Provider
          mockInitialState={{ companiesInitial: companies, companies }}
        >
          <MyCompaniesTile />
        </useMyCompaniesContext.Provider>,
      )
      wrapper.find('input').simulate('change', { target: { value: 'z' } })
      expect(wrapper.find('table').text()).toContain('Zebra')
      expect(wrapper.find('tbody')).toHaveLength(1)
    })
  })
  describe('when text filter doesn\'nt match any companies', () => {
    const companiesInitial = [
      {
        company: {
          name: 'A1 BMW LTD!!!!!',
        },
        latestInteraction: {
          id: '79d92719-7402-45b6-b3d7-eff559d6b282',
          date: '2019-08-14',
          displayDate: '14 Aug 19',
          subject:
            'Here is a long interaction title some more text some more text some more text almost finished some more text nearly there more text finished',
        },
      },
    ]
    const wrapper = mount(
      <useMyCompaniesContext.Provider mockInitialState={{ companiesInitial, filterText: 'Z1', companies: [] }}>
        <MyCompaniesTile />
      </useMyCompaniesContext.Provider>,
    )

    test('should render the no companies message', () => {
      expect(wrapper.find('p').text()).toContain('No companies match your search')
    })
  })
})
