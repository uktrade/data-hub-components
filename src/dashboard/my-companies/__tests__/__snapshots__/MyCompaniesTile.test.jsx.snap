// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My companies dashboard should match the snapshot 1`] = `
.c3 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  display: block;
  margin-bottom: 15px;
  color: #6f777b;
}

.c2 {
  color: #0b0c0c;
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 700;
  font-size: 24px;
  line-height: 1.0416666666666667;
  display: block;
  margin-top: 0;
  margin-bottom: 20px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.c1 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

@media print {
  .c3 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c3 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c2 {
    color: #000;
  }
}

@media print {
  .c2 {
    font-size: 24px;
    line-height: 1.05;
  }
}

@media only screen and (min-width:641px) {
  .c2 {
    font-size: 36px;
    line-height: 1.1111111111111112;
  }
}

@media only screen and (min-width:641px) {
  .c2 {
    margin-bottom: 30px;
  }
}

<Provider
  mockInitialState={
    Object {
      "companies": Array [
        Object {
          "company": Object {
            "id": "a1138c1c-d449-4846-aa58-18fae7e1cb92",
            "isArchived": false,
            "name": "A1 BMW LTD!!!!!",
          },
          "latestInteraction": Object {
            "date": "2019-08-14",
            "displayDate": "14 Aug 19",
            "id": "79d92719-7402-45b6-b3d7-eff559d6b282",
            "subject": "Here is a long interaction title some more text some more text some more text almost finished some more text nearly there more text finished",
          },
        },
        Object {
          "company": Object {
            "id": "a30b1db3-7140-44ca-ad7a-9824c3c2ed56",
            "isArchived": false,
            "name": "Portable Potatoes",
          },
          "latestInteraction": Object {
            "date": null,
            "displayDate": "-",
            "subject": "No interactions have been recorded",
          },
        },
        Object {
          "company": Object {
            "id": "b89b1db3-7140-44ca-ad7a-9824c3c2gh74",
            "isArchived": false,
            "name": "Zebra clothing",
          },
          "latestInteraction": Object {
            "date": "2019-02-21",
            "displayDate": "21 Feb 2019",
            "id": "86f92719-7402-45b6-b3d7-eff559d6b678",
            "subject": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
          },
        },
      ],
      "companiesInitial": Array [
        Object {
          "company": Object {
            "id": "a1138c1c-d449-4846-aa58-18fae7e1cb92",
            "isArchived": false,
            "name": "A1 BMW LTD!!!!!",
          },
          "latestInteraction": Object {
            "date": "2019-08-14",
            "displayDate": "14 Aug 19",
            "id": "79d92719-7402-45b6-b3d7-eff559d6b282",
            "subject": "Here is a long interaction title some more text some more text some more text almost finished some more text nearly there more text finished",
          },
        },
        Object {
          "company": Object {
            "id": "a30b1db3-7140-44ca-ad7a-9824c3c2ed56",
            "isArchived": false,
            "name": "Portable Potatoes",
          },
          "latestInteraction": Object {
            "date": null,
            "displayDate": "-",
            "subject": "No interactions have been recorded",
          },
        },
        Object {
          "company": Object {
            "id": "b89b1db3-7140-44ca-ad7a-9824c3c2gh74",
            "isArchived": false,
            "name": "Zebra clothing",
          },
          "latestInteraction": Object {
            "date": "2019-02-21",
            "displayDate": "21 Feb 2019",
            "id": "86f92719-7402-45b6-b3d7-eff559d6b678",
            "subject": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
          },
        },
      ],
    }
  }
>
  <MyCompaniesTile>
    <div>
      <ListSelector>
        <styled.div>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-kGXeez",
                  "isStatic": true,
                  "lastClassName": "c0",
                  "rules": Array [
                    "
  display: flex;
  align-items: baseline;
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-kGXeez",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <div
              className="c0"
            >
              <Styled(H2)>
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-kpOJdX",
                        "isStatic": true,
                        "lastClassName": "c1",
                        "rules": Array [
                          "
  flex-grow: 1;
",
                        ],
                      },
                      "displayName": "Styled(H2)",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-kpOJdX",
                      "target": [Function],
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <H2
                    className="c1"
                  >
                    <Heading
                      as="h2"
                      className="c1"
                      size="LARGE"
                    >
                      <styled.h1
                        as="h2"
                        className="c1"
                        size="LARGE"
                      >
                        <StyledComponent
                          as="h2"
                          className="c1"
                          forwardedComponent={
                            Object {
                              "$$typeof": Symbol(react.forward_ref),
                              "attrs": Array [],
                              "componentStyle": ComponentStyle {
                                "componentId": "sc-kgoBCf",
                                "isStatic": false,
                                "lastClassName": "c2",
                                "rules": Array [
                                  "color: #0b0c0c; @media print {
  color: #000;
}",
                                  [Function],
                                  "display: block; margin-top: 0;",
                                  [Function],
                                  [Function],
                                ],
                              },
                              "displayName": "styled.h1",
                              "foldedComponentIds": Array [],
                              "render": [Function],
                              "styledComponentId": "sc-kgoBCf",
                              "target": "h1",
                              "toString": [Function],
                              "warnTooManyClasses": [Function],
                              "withComponent": [Function],
                            }
                          }
                          forwardedRef={null}
                          size="LARGE"
                        >
                          <h2
                            className="c1 c2"
                            size="LARGE"
                          >
                            My companies list
                          </h2>
                        </StyledComponent>
                      </styled.h1>
                    </Heading>
                  </H2>
                </StyledComponent>
              </Styled(H2)>
            </div>
          </StyledComponent>
        </styled.div>
      </ListSelector>
      <NoListsMsg>
        <HintText>
          <styled.span>
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bdVaJa",
                    "isStatic": false,
                    "lastClassName": "c3",
                    "rules": Array [
                      "font-family: \\"nta\\", Arial, sans-serif; -webkit-font-smoothing: antialiased; -moz-osx-font-smoothing: grayscale; @media print {
  font-size: 14px; line-height: 1.15;
} font-weight: 400; font-size: 16px; line-height: 1.25; @media only screen and (min-width: 641px) {
  font-size: 19px; line-height: 1.3157894736842106;
}",
                      "display: block; margin-bottom: 15px; color: #6f777b;",
                      [Function],
                    ],
                  },
                  "displayName": "styled.span",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-bdVaJa",
                  "target": "span",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
            >
              <span
                className="c3"
              >
                You have not yet created any lists with companies.
                <br />
                You can add companies to lists from a company page, and only you can see these lists.
              </span>
            </StyledComponent>
          </styled.span>
        </HintText>
      </NoListsMsg>
    </div>
  </MyCompaniesTile>
</Provider>
`;
